go.mod

```diff
@@ -9,6 +9,7 @@ go 1.13
 require (
 	code.cloudfoundry.org/bytefmt v0.0.0-20190710193110-1eb035ffe2b6
 	github.com/PuerkitoBio/goquery v1.5.0
+	github.com/appscode/jsonpatch v1.0.1
```

Comment:
I would suggest using `github.com/evanphx/json-patch` which also used by kubernetes repo. For example https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/cmd/kubeadm/app/util/patches/patches.go#L31.

Reply:
Also could avoid potential conflicting with kubernetes repo in the future

Reply:
I cannot find any methods that could create a JSON path after go through `github.com/evanphx/json-patch`. It looks like only providing the way to apply a patch.

Reply:
OK, I see what's going on here. You were trying to solve the issue `Operation cannot be fulfilled on pipelines.devops.kubesphere.io "pipe2": the object has been modified; please apply your changes to the latest version and try again` when updating an object. This happens when updating with resourceVersion in the body, if still want to update even the object has been modified by other clients, I think you'd better update without resourceVersion. Patching makes things more complicated.

Reply:
The `ResourceVersion` cannot be empty. Because it'll cause the error: `metadata.resourceVersion: Invalid value: 0x0: must be specified for an update`. Or do you have any good examples to achieve that?

Reply:
As @leoendless said, get resourceVersion before updating would help decrease these errors. I don't think patching is the best solution here.

Reply:
It can reduce the chance of this kind of errors happen. But if you believe that we should take this solution instead of the patch. I'd prefer to create a new PR. In my view, get the latest resourceVersion then update it as some kind of patch.

see also #3349